Add `as_uninit`-like methods to pointer types and unify documentation of `as_ref` methods

TimDiekmann opened the pull request:
This adds a convenient method to retrieve a `&(mut) [MaybeUninit<T>]` from slice pointers (`*const [T]`, `*mut [T]`, `NonNull<[T]>`). See also https://github.com/rust-lang/wg-allocators/issues/66#issuecomment-671789105.

~I'll add a tracking issue as soon as it's reviewed and CI passed.~
Tracking Issue: #75402

r? @RalfJung

RalfJung commented:
Is there any particular reason why you added this for slices, but not for individual elements? It seems just as useful for `*const T`/`*mut T`/`NonNull<T>`.

RalfJung commented on library/core/src/ptr/mut_ptr.rs:
```diff
@@ -1112,6 +1112,42 @@ impl<T> *mut [T] {
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * The pointer must be [valid] for reads and writes for `ptr.len() * mem::size_of::<T>()` many bytes,
+    ///   and it must be properly aligned. This means in particular:
```
```suggestion
    ///     and it must be properly aligned. This means in particular:
```

TimDiekmann replied:
I'm sorry, I don't understand what you mean. Are you talking about the term "properly aligned" or the indentation of the commentary? According to your edit, you meant the latter, but I don't get it.

RalfJung replied:
Ah lol "alignment" is ambiguous. ;)

I meant the indention. But I just realized you probably deliberately made these next two bullets further indented as sub-bullets. Let's hope this renders correctly, which I find really hard to predict with markdown.

TimDiekmann replied:
The comment is largely copied from `slice::from_raw_parts(_mut)`, so it should work :)

TimDiekmann commented:
That would would Result in `MaybeUninit<[T]>` rather than `[MaybeUninit<T>]`.

TimDiekmann commented:
Maybe we could also add `as_uninit` for non-slices?

RalfJung commented:
> That would would result in MaybeUninit<[T]> rather than [MaybeUninit<T>].

We probably want a method that converts between those two as well, but I don't know a good way to add such a method. :/

> Maybe we could also add as_uninit for non-slices?

That's what I meant -- something like `as_uninit`/`as_uninit_ref` or so.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -906,6 +906,41 @@ impl<T> *const [T] {
+    /// [`pointer::offset`]: ../primitive.pointer.html#method.offset
+    #[inline]
+    #[unstable(feature = "uninit_ptr_slice", issue = "none")]
+    pub unsafe fn as_uninit_slice<'a>(self) -> &'a [MaybeUninit<T>] {
```
Please note in the tracking issue that the naming is different from the usual raw ptr methods which indicate mutability -- like `from_raw_parts_mut` or `as_mut`. I don't have a strong opinion either way, but the libs team might. ;)

TimDiekmann replied:
I could rename the feature to `slice_ptr_uninit` which follows the naming scheme of #74265.

RalfJung replied:
I don't think anyone cares about feature names, but sure.^^
I mean the names of the methods.

TimDiekmann replied:
Oh sure :smile: 

Well, currently the term "uninit" is always prepended to the type: https://doc.rust-lang.org/nightly/core/index.html?search=uninit

RalfJung replied:
I was talking about the fact that the const and mut methods have the same name, which they usually do not.

TimDiekmann replied:
Oh :D
I'll change this.

TimDiekmann commented:
> We probably want a method that converts between those two as well, but I don't know a good way to add such a method. :/

Me neither, we probably want a method converting in both direction with a decent name...

> That's what I meant -- something like `as_uninit`/`as_uninit_ref` or so.

I'd name them `as_uninit_ref` to make them analogue to `as_ref`. Do we want another pull request for this or should I just add this here, gated behind `as_uninit_ref`? Should I add both to the same tracking issue?

RalfJung commented:
I'd make them all one PR and one feature.

TimDiekmann commented:
Currently we have `NonNull::as_ref` and `NonNull::as_mut` taking `&(mut) self` (both stabilized - sadly). Shall I also alter `as_uninit_slice` to take `&self` and returning `&[MaybeUninit]` and add `as_uninit_slice_mut`? Since `NonNull` implements `Copy` I don't like passing it as a reference, as you will almost never have a reference to `NonNull`, especially not a `&mut NonNull`.

This btw. also applies to `as_mut_ptr` which converts a slice to a non-slice pointer. On nightly there is `MaybeUninit::first_ptr(_mut)` which maybe could also be used on slice pointers instead. With this naming we could keep a single naming scheme.

RalfJung commented:
> Shall I also alter as_uninit_slice to take &self and returning &[MaybeUninit] and add as_uninit_slice_mut?

I can do technical reviews but this is a question for @rust-lang/libs that I cannot help with, I am afraid.

TimDiekmann commented:
I'll first implement the methods as they were already implemented with the pointer types. Since some methods are already stable, I think that a unified schema is the most important thing.

TimDiekmann commented:
I just noticed, that `pointer::as_ref` returns `Option<&T>`.

TimDiekmann commented:
> Is there any particular reason why you added this for slices, but not for general pointers? It seems just as useful for `*const T`/`*mut T`/`NonNull<T>`.

One thing to note: `MaybeUninit<[T]>` isn't even possible right now, as `MaybeUninit<T>` requires `T` to be sized.
Noticed that when I implemented `as_uninit_ref/mut` :D

RalfJung commented:
Ah good point, and we are unlikely to be able to relax that sized bound.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -101,6 +100,35 @@ impl<T: ?Sized> *const T {
+    ///
+    /// While this method is useful for null-safety, it is important to note
+    /// that this is still an unsafe operation because the returned value could
+    /// be pointing to invalid memory.
```
"Invalid memory" in the sense of "containing invalid data" is not a concern.

But the pointer could be a non-null dangling pointer. That would be a problem.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -101,6 +100,35 @@ impl<T: ?Sized> *const T {
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// it is properly aligned
```
It must also be dereferencable.

RalfJung commented on library/core/src/ptr/mut_ptr.rs:
```diff
@@ -100,6 +100,36 @@ impl<T: ?Sized> *mut T {
+    /// memory.
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// it is properly aligned
```
Likewise, it must also be dereferencable.

RalfJung commented on library/core/src/ptr/mut_ptr.rs:
```diff
@@ -100,6 +100,36 @@ impl<T: ?Sized> *mut T {
+    ///
+    /// # Safety
+    ///
+    /// While this method and its mutable counterpart are useful for
```
Above you removed the "and its counterpart", I think it would make sense to also remove it here (and in `as_ref`). Alternatively, link to what that counterpart is.

RalfJung commented on library/core/src/ptr/mut_ptr.rs:
```diff
@@ -286,6 +316,37 @@ impl<T: ?Sized> *mut T {
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// it is properly aligned
```
Likewise it must be dereferencable.

RalfJung commented on library/core/src/ptr/non_null.rs:
```diff
@@ -76,6 +76,54 @@ impl<T: Sized> NonNull<T> {
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that `self` is properly aligned.
```
and it must be dereferencable

RalfJung commented on library/core/src/ptr/non_null.rs:
```diff
@@ -76,6 +76,54 @@ impl<T: Sized> NonNull<T> {
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that `self` is properly aligned.
```
and it must be dereferencable

RalfJung commented on library/core/src/ptr/mut_ptr.rs:
```diff
@@ -105,20 +106,21 @@ impl<T: ?Sized> *mut T {
     /// # Safety
     ///
-    /// While this method and its mutable counterpart are useful for
+    /// While this method and its [mutable counterpart] are useful for
```
For symmetry I would then expect the same wording on that mutable counterpart -- something like "this method and its shared counterpart". Doesn't seem like that's what we got though? (It's a bit hard to tell in the GH diff.)

TimDiekmann commented:
Those intra-doc links drive me crazy :D
CI should pass now.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -906,6 +933,50 @@ impl<T> *const [T] {
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
+    ///   and it must be properly aligned. This means in particular:
```
The requirement for `as_uninit_ref` and `as_uninit_slice` is basically the same -- aligned and dereferencable -- but it is stated very differently. I am not sure what is better, but this is a curious inconsistency.

TimDiekmann replied:
I personally like the wording of `as_uninit_slice` better. It is easier to skim a list of bullet points than to read a coherent text.

RalfJung replied:
Agreed. That's also how [`ptr:.write`](https://doc.rust-lang.org/nightly/std/ptr/fn.write.html) and similar methods look.

TimDiekmann replied:
Do you want me to change the docs for `as_ref` and `as_mut`, too? Then they are unified.

RalfJung replied:
That would be great if you could do that. :)

TimDiekmann replied:
I think I have everything added. Please review the safety clauses carefully, let's don't risk I forgot something :)

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -38,32 +38,30 @@ impl<T: ?Sized> *const T {
+    ///
+    /// * it must be "dereferencable" in the sense defined in [the module documentation], and
+    ///
+    /// * the lifetime `'a` returned is arbitrarily chosen and does
```
The wording is odd. You must ensure that the lifetime is arbitrarily chosen? I'd start with something like "*You* must enforce Rust's aliasing rules, since the returned lifetime `'a` is arbitrarily chosen. In particular, ..."

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -38,32 +38,30 @@ impl<T: ?Sized> *const T {
+    ///   Rust's aliasing rules. In particular, for the duration of this lifetime,
+    ///   the memory the pointer points to must not get mutated (except inside `UnsafeCell`).
+    ///
+    /// * the pointer must point to an initialized instance of `T`.
```
I'd move this up above the lifetime.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -101,6 +99,70 @@ impl<T: ?Sized> *const T {
+    /// let ptr: *const u8 = &10u8 as *const u8;
+    ///
+    /// unsafe {
+    ///     let val_back = &*(ptr as *const MaybeUninit<u8>);
```
I am always hesitant to recommend pointer casts. They are a very dangerous operation as the type checker checks very little about them.

There's no unwrap_unchecked, is there?

TBH I never understood why NULL was treated separately from the other erorr conditions in these methods, but I expect it is too confusing to make the uninit versions return a plain reference without `Option`?

TimDiekmann replied:
I just copied the previous example but I don't like the example either. Actually the method is unsafe anyway, so dereferencing a raw pointer in this context seems ok'ish.

No, sadly, there is no `unwrap_unchecked` but there should be IMO. If we add it, this example should definitely be changed.

Yes, that was the reason about it. I'd expect `as_uninit_ref` to behave exactly as `as_ref`, besides wrapped in an `MaybeUninit`. Also it is UB to have `&MaybeUninit<T>` pointing to NULL. For a pointer guaranteed to be non-NULL we have `NonNull`, which should be used whenever possible IMO.

RalfJung replied:
The example you copied doesn't need a pointer cast though.

TimDiekmann replied:
Oh, I'm sorry. Maybe we should just leave that out then?

RalfJung replied:
It's still a fair point though, there should be an ergonomic way to do that, and IMO "ergonomic" implies "no `as`".

I am fine with punting on that for now and leaving a note in the tracking issue.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -38,32 +38,30 @@ impl<T: ?Sized> *const T {
+    ///
+    /// * The pointer must be properly aligned,
+    ///
+    /// * it must be "dereferencable" in the sense defined in [the module documentation], and
```
You have a ", and" here but not on the first item. Given that the list starts with "all of the following must be true", I don't think the ", and" is necessary, but if you want to use it please do so consistently.

TimDiekmann replied:
The " and" was a copy-paste misstake, thanks!

RalfJung replied:
Basically these should be either self-contained sentences (starting with a capital letter and ending in a `.`), or forma  long sentence together. But the last gets kind of awkward given that the last item is multiple sentences long.

TimDiekmann replied:
Is there a guideline for writing safety-sections somewhere? It seems that there are several approaches, even in `std`. A few examples:
- [`str::get_unchecked`](https://doc.rust-lang.org/nightly/std/primitive.str.html#method.get_unchecked) uses ";" and "."
- [`slice::from_raw_parts`](https://doc.rust-lang.org/nightly/core/slice/fn.from_raw_parts.html) or [`str::slice_unchecked`](https://doc.rust-lang.org/nightly/std/primitive.str.html#method.slice_unchecked) uses "."
- [`GlobalAlloc::realloc`](https://doc.rust-lang.org/nightly/core/alloc/trait.GlobalAlloc.html#method.realloc) (and other methods from `core::alloc`) uses a mix of ", and", ",", and "."

RalfJung replied:
I don't think there is... I'd have used `.` for all of those.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -101,6 +99,53 @@ impl<T: ?Sized> *const T {
     }
 
+    /// Returns `None` if the pointer is null, or else returns a shared reference to
+    /// the value wrapped in `Some`.
```
This should probably mention "uninitialized" somewhere. It could point out explicitly that the key difference to `as_ref`. is that the pointed-to data does *not* have to be initialized.

TimDiekmann replied:
May I add something like "In comparison to [`as_ref`](link to as_ref), this does not require, that the value has to be initialized."?

RalfJung replied:
I'd probably say "in contrast to", and the last "," doesn't belong there I think, but yes. :)

TimDiekmann replied:
Alright. Well, the last "," is German grammar I think 😄

TimDiekmann replied:
I have also added a link from `as_ref` to `as_uninit_ref` to quickly jump back.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -906,6 +951,52 @@ impl<T> *const [T] {
+    ///       for zero-length slices using [`NonNull::dangling()`].
+    ///
+    /// * The total size `ptr.len() * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`.
+    ///   See the safety documentation of [`pointer::offset`],
```
Ending this in a "," even though this is a two-sentence bullet is somewhat odd.

RalfJung commented:
In terms of text this is all fine (assuming it's the same text for correspond methods, which I did not check in detail everywhere). I just have some more typographic nits.

TimDiekmann commented:
Thank you!
I'll double-check every text and it's method again and will push it later this day.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -966,16 +975,16 @@ impl<T> *const [T] {
     ///       Slices can never span across multiple allocated objects, and
     ///
-    ///     * The pointer must be aligned even for zero-length slices. One
+    ///     * the pointer must be aligned even for zero-length slices. One
```
"Slices can never span across multiple allocated objects, and the pointer must be aligned even for zero-length slices." -- that's an odd sentence, so I think this is worth changing to proper sentences as well.

(I hope I am not nitpicking too much here, if you want we can also land this and do a cleanup/editing PR afterwards.)

TimDiekmann replied:
That review is already outdated ;)

RalfJung replied:
Ah, I got no further email from GH but I guess you force-pushed?

TimDiekmann replied:
No, I think your diff was outdated. The changes where pushed with 82b8d75 :)
I try not to force push whenever possible.

RalfJung commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -906,6 +957,55 @@ impl<T> *const [T] {
+    ///   and it must be properly aligned. This means in particular:
+    ///
+    ///     * The entire memory range of this slice must be contained within a single allocated object!
+    ///       Slices can never span across multiple allocated objects, and
+    ///
+    ///     * the pointer must be aligned even for zero-length slices. One
```
Still looks like such an odd sentence here though?

TimDiekmann replied:
For me it felt natural to continue the sentence there, since both complement the previous sentence, but I think that's a matter of taste. If you want, I can change that.

RalfJung replied:
The previous sentence here is "The entire memory range of this slice must be contained within a single allocated object!". I don't see how the second bullet point complements that point.

The two bullets are two separate independent points, and yet somehow the second half of the first bullet and the beginning of the second bullet are glued into one sentence. I don't see how that makes any sense.

RalfJung replied:
Maybe you mean that both complement "The pointer must be valid for reads", but that is not what the grammatical structure indicates here. The sentence structure indicates a connection between "must be contained within a single allocated object" on the one hand, and ("never span across multiple allocated objects" + "must be aligned even for zero-length slices") on the other hand, which to me just makes no sense.

TimDiekmann replied:
Well, I think you are right. Reinvestigating [`slice::from_raw_parts`](https://doc.rust-lang.org/nightly/core/slice/fn.from_raw_parts.html) shows, that its documentation also uses two separate sentences.

TimDiekmann replied:
Do you think it is enough if I change it as follows, or should it be wound up a bit more?
```suggestion
    ///       Slices can never span across multiple allocated objects.
    ///
    ///     * The pointer must be aligned even for zero-length slices. One
```

RalfJung replied:
That diff seems fine to me. What other changes were you thinking of?

TimDiekmann replied:
I do not have any concrete ones in mind at the moment, maybe you can think of something else?

RalfJung replied:
Not really, that's why I am asking. ;)
Let's just go with that diff.

RalfJung commented on library/core/src/ptr/non_null.rs:
```diff
@@ -278,6 +352,114 @@ impl<T> NonNull<[T]> {
+    /// use std::ptr::NonNull;
+    ///
+    /// let memory: NonNull<[u8]> = Global.alloc(Layout::new::<[u8; 32]>())?;
+    /// // This is safe as `memory` is a valid pointer to `[u8]`
```
It's not initialized though, so it's not really valid?

TimDiekmann replied:
From `core::ptr#Safety`:

> For a pointer to be valid, it is necessary, but not always sufficient, that the pointer be dereferenceable: the memory range of the given size starting at the pointer must all be within the bounds of a single allocated object.

The safety constraint from `AllocRef` requires this:
> Memory blocks returned from an allocator must point to valid memory.

We have to elaborate on this in the allocator-wg, but this is another topic. Basically it is required for all allocators, that a valid pointer is returned. For this, there are currently two issues open: https://github.com/rust-lang/wg-allocators/issues/65 https://github.com/rust-lang/wg-allocators/issues/46. The former wants to refine the quoted phrase from `AllocRef`.

To sum up: This is not initialized, but valid. As long as we don't return `&[T]` this should be fine.

, that it's dereferencable. An allocator returns a well-aligned pointer to an allocated object. The pointer does not necessarily need to point to initialized memory for this method to be called. I think this is the whole point of `MaybeUninit`.

RalfJung replied:
That is *valid for a given access kind and size*. But what your text here says sounds more like "valid for a given type", which -- for me at least -- is something very different.

TimDiekmann replied:
I don't think I quite follow. Which validity invariant am I missing here?

The whole reason to use an allocator (and let's assume that `Global` is implemented correctly here) is to get back a memory region that we can write to. Such a memory area may contain anything, uninitialized, but has a fixed size of at least `NonNull::len()`. For me this is just like returning a slice of `MaybeUninit<u8>` with a size of `len()`. Since `alloc` returns `NonNull<[u8]>`, it is guaranteed that no padding exists between `ptr[0]` and `ptr[1]` as [`u8` has no padding bits and its size is exactly one byte](https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#fixed-width-integer-types). So why is `&[MaybeUninit<u8>]` not a valid representation?

It may well be that I overlook or misunderstand something, but at the moment I cannot see the reason.

RalfJung replied:
> Which validity invariant am I missing here?

`u8` must be initialized. So allocators do not return valid `u8`.

Your safety docs get it right, it says

> The pointer must be valid for reads (and writes) for `ptr.len() * mem::size_of::<T>()` many bytes

So the code's safety comment should match those docs. But they do not.
The correct thing to say would be something like "`memory` is valid for read and writes for `memory.len()` many bytes".

TimDiekmann replied:
Oh, I'm so glad. I was a little shocked :D

Sure, the comment is very poor worded and "valid" isn't the right term here. I think your suggestion makes it more clear:

```suggestion
    /// // This is safe as `memory` is a valid for reads and writes for `memory.len()` many bytes.
    /// // Note, that calling `memory.as_mut()` is not allowed here as the content may be uninitialized.
```

RalfJung replied:
"is a valid" -> "is valid"

Also I don't think there should be a comma after "Note".

RalfJung commented on library/core/src/ptr/non_null.rs:
```diff
@@ -278,6 +352,114 @@ impl<T> NonNull<[T]> {
+    ///
+    /// When calling this method, you have to ensure that all of the following is true:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
```
"reads" and writes?

TimDiekmann replied:
Oh, yes 😄

RalfJung commented on library/core/src/ptr/mut_ptr.rs:
```diff
@@ -1112,6 +1211,110 @@ impl<T> *mut [T] {
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// all of the following is true:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
```
and writes

RalfJung commented:
All right, I think this is ready. :) Could you squash the commits?

TimDiekmann committed:
Add `as_uninit`-like methods to pointer types and unify documentation of `as_ref` methods

Fix example in `NonNull::as_uninit_slice`


Rename feature gate to "ptr_as_uninit"


Make methods more consistent with already stable methods


Make `pointer::as_uninit_slice` return an `Option`


Fix placement for `// SAFETY` section


Add `as_uninit_ref` and `as_uninit_mut` to pointers


Fix doctest


Update tracking issue


Fix doc links


Apply suggestions from review


Make wording about counterparts consistent


Fix doc links


Improve documentation

Fix doc-tests


Fix doc links ... again


Apply suggestions from review


Apply suggestions from Review


Apply suggestion from review to all affected files


Add missing words in safety sections in `as_uninit_slice_mut`


Fix safety-comment in `NonNull::as_uninit_slice_mut`

TimDiekmann commented:
Done (in case GH does not send mails on force-pushes)

RalfJung commented:
> Done (in case GH does not send mails on force-pushes).

(it doesn't, except sometimes when it goes... their support told me it only sends them when the parent of the new HEAD was already in the branch, aka when you just amended the last commit.)

Great, and thanks for enduring by never-ending stream of editorial comments. :)
@bors r+ rollup

bors commented:
:pushpin: Commit 93e074bc8a0766afbe3594f8663d702638ec0c35 has been approved by `RalfJung`

<!-- @bors r=RalfJung 93e074bc8a0766afbe3594f8663d702638ec0c35 -->
<!-- homu: {"type":"Approved","sha":"93e074bc8a0766afbe3594f8663d702638ec0c35","approver":"RalfJung"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #75592

bors commented:
:pushpin: Commit 93e074bc8a0766afbe3594f8663d702638ec0c35 has been approved by `RalfJung`

<!-- @bors r=RalfJung 93e074bc8a0766afbe3594f8663d702638ec0c35 -->
<!-- homu: {"type":"Approved","sha":"93e074bc8a0766afbe3594f8663d702638ec0c35","approver":"RalfJung"} -->

TimDiekmann commented:
> Great, and thanks for enduring by never-ending stream of editorial comments. :)

I think the effort was worth it :)
